Add CTCI 2.2 (Kth to Last) for Linked List session (#95)
UWB-ACM/CTCI: 2019-03-Autumn/1_linked_lists/README.md

#### Iterative Solution (Simple)

If we know that we desire the `k`th to last element of the list, we can 
easily displace two separate pointers `k - 1` distance apart from each 
other. When these pointers move in lockstep down the list and the 
`leader` pointer arrives at the last element of the list, we know that 
the `follower` has arrived at the desired element.

A Python implementation would be:

```python3
def get_kth_to_last_iterative(k: int, head: Node = None):
    if head is None:
        return None
    leader = head
    target = head
    # Move our leader pointer forward k - 1 nodes.
    for x in range(k - 1):
        # Edge case: our list is shorter than k nodes.
        if leader.next is None:
            return None
        else:
            leader = leader.next
    # Move leader and target pointers forward in lockstep until 
    # the end of the list is reached.
    while leader.next is not None:
        leader = leader.next
        target = target.next
    return target
```

#### Recursive Solution (Less Simple)

Achieving the same result using recursion is more cumbersome when 
compared to the iterative solution and requires some inventiveness on 
our part.

> In theory, we should be able to traverse the entire list and increment 

Author: ```suggestion
In theory, we should be able to traverse the entire list and then increment 
```

> a counter as the stack frames resolve until the counter reaches `k`. 

Author: ```suggestion
a counter as each recursive call returns until the counter reaches `k`. 
```

> However, this presents several problems for implementation: 

Reviewer: stack frames 'resolve' kinda sounds weird

Author: Thank ye

and specific programming languages.

We present a Python solution which uses language-specific features to 
achieve the objective.

```python3
def get_kth_to_last_recursive(k: int, head: Node = None):
    # edge case
    if head is None:
        return None
    # enter recursive function, return desired value
    return recurse_kth(k, head, head)[1]

def recurse_kth(k: int, current: Node, head: Node):
    # Base case: we have reached end of list.
    # Return the value we expect to report for the previous node, 
    # along with a boolean flag about whether our search should continue.
    if current is None:
        return True, 1
    proceed, depth = recurse_kth(k, current.next, head)
    # We have terminated; pass the reported values up the stack.
    if not proceed:
        return False, depth
    # We have found our target node.
    if depth == k:
        return False, current
    # We need to continue looking for our target node.
    elif depth < k and current is not head:
        return True, depth + 1
    # If no previous conditions are met, we will not be able to find a match.
    return False, None
```

> #### Testing The Solutions OR Driver For Solution

Author: ```suggestion
#### Testing The Solutions
```
